File: db4o.net/Db4objects.Db4o.Tests/Db4objects.Db4o.Tests/Common/Concurrency/DeleteDeepTestCase.cs
/* Copyright (C) 2004 - 2011  Versant Inc.  http://www.db4o.com */

#if !SILVERLIGHT
using Db4oUnit;
using Db4oUnit.Extensions;
using Db4objects.Db4o;
using Db4objects.Db4o.Config;
using Db4objects.Db4o.Ext;
using Db4objects.Db4o.Query;
using Db4objects.Db4o.Tests.Common.Concurrency;

namespace Db4objects.Db4o.Tests.Common.Concurrency
{
	public class DeleteDeepTestCase : Db4oClientServerTestCase
	{
		public static void Main(string[] args)
		{
			new DeleteDeepTestCase().RunConcurrency();
		}

		public string name;

		public DeleteDeepTestCase child;

		protected override void Store()
		{
			AddNodes(10);
			name = "root";
			Store(this);
		}

		protected override void Configure(IConfiguration config)
		{
			config.ObjectClass(typeof(DeleteDeepTestCase)).CascadeOnDelete(true);
		}

		// config.objectClass(DeleteDeepTestCase.class).cascadeOnActivate(true);
		private void AddNodes(int count)
		{
			if (count > 0)
			{
				child = new DeleteDeepTestCase();
				child.name = string.Empty + count;
				child.AddNodes(count - 1);
			}
		}

		/// <exception cref="System.Exception"></exception>
		public virtual void Conc(IExtObjectContainer oc)
		{
			IQuery q = oc.Query();
			q.Constrain(typeof(DeleteDeepTestCase));
			q.Descend("name").Constrain("root");
			IObjectSet os = q.Execute();
			if (os.Count == 0)
			{
				// already deleted
				return;
			}
			Assert.AreEqual(1, os.Count);
			if (!os.HasNext())
			{
				return;
			}
			DeleteDeepTestCase root = (DeleteDeepTestCase)os.Next();
			// wait for other threads
			// Thread.sleep(500);
			oc.Delete(root);
			oc.Commit();
			AssertOccurrences(oc, typeof(DeleteDeepTestCase), 0);
		}

		public virtual void Check(IExtObjectContainer oc)
		{
			AssertOccurrences(oc, typeof(DeleteDeepTestCase), 0);
		}
	}
}
#endif // !SILVERLIGHT

File: db4oj.tests/src/com/db4o/db4ounit/common/concurrency/DeleteDeepTestCase.java
/* Copyright (C) 2004 - 2007  Versant Inc.   http://www.db4o.com */
package com.db4o.db4ounit.common.concurrency;

import com.db4o.*;
import com.db4o.config.*;
import com.db4o.ext.*;
import com.db4o.query.*;

import db4ounit.*;
import db4ounit.extensions.*;

public class DeleteDeepTestCase extends Db4oClientServerTestCase {

	public static void main(String[] args) {
		new DeleteDeepTestCase().runConcurrency();
	}

	public String name;

	public DeleteDeepTestCase child;

	protected void store() {
		addNodes(10);
		name = "root";
		store(this);
	}

	protected void configure(Configuration config) {
		config.objectClass(DeleteDeepTestCase.class).cascadeOnDelete(true);
		// config.objectClass(DeleteDeepTestCase.class).cascadeOnActivate(true);
	}

	private void addNodes(int count) {
		if (count > 0) {
			child = new DeleteDeepTestCase();
			child.name = "" + count;
			child.addNodes(count - 1);
		}
	}

	public void conc(ExtObjectContainer oc) throws Exception {
		Query q = oc.query();
		q.constrain(DeleteDeepTestCase.class);
		q.descend("name").constrain("root");
		ObjectSet os = q.execute();
		if (os.size() == 0) { // already deleted
			return;
		}
		Assert.areEqual(1, os.size());
		if(!os.hasNext()){
			return;
		}
		DeleteDeepTestCase root = (DeleteDeepTestCase) os.next();
		
		// wait for other threads
		// Thread.sleep(500);
		oc.delete(root);
		oc.commit();
		assertOccurrences(oc, DeleteDeepTestCase.class, 0);
	}

	public void check(ExtObjectContainer oc) {
		assertOccurrences(oc, DeleteDeepTestCase.class, 0);
	}

}
